packages/rest/src/controllers/didcomm/credentials/CredentialsController.ts

```diff
@@ -0,0 +1,281 @@
+  AcceptCredentialOfferOptions,
+  CreateOfferOptions,
+  credentialExchangeRecordToApiModel,
+} from './CredentialsControllerTypes'
+
+@Tags('DIDComm Credentials')
+@Route('/didcomm/credentials')
+@Security('tenants', ['tenant'])
+@injectable()
+export class CredentialsController extends Controller {
+  /**
+   * Retrieve all credential exchange records by query
+   */
+  @Example<DidCommCredentialsExchangeRecord[]>([credentialExchangeRecordExample])
+  @Get('/')
+  public async findCredentialsByQuery(
+    @Request() request: RequestWithAgent,
+    @Query('threadId') threadId?: ThreadId,
+    @Query('parentThreadId') parentThreadId?: ThreadId,
+    @Query('connectionId') connectionId?: RecordId,
+    @Query('state') state?: CredentialState,
+    @Query('role') role?: CredentialRole,
+  ) {
+    const credentials = await request.user.agent.credentials.findAllByQuery({
+      connectionId,
+      threadId,
+      state,
+      parentThreadId,
+      role,
+    })
+
+    return credentials.map(credentialExchangeRecordToApiModel)
+  }
+
+  /**
+   * Retrieve credential exchange record by credential record id
+   *
+   * @param credentialExchangeId
+   * @returns CredentialExchangeRecord
+   */
+  @Example<DidCommCredentialsExchangeRecord>(credentialExchangeRecordExample)
+  @Get('/:credentialExchangeId')
+  public async getCredentialById(
+    @Request() request: RequestWithAgent,
+    @Path('credentialExchangeId') credentialExchangeId: RecordId,
+  ) {
+    try {
+      const credential = await request.user.agent.credentials.getById(credentialExchangeId)
+      return credentialExchangeRecordToApiModel(credential)
+    } catch (error) {
+      if (error instanceof RecordNotFoundError) {
+        this.setStatus(404)
+        return apiErrorResponse(`credential exchange with id "${credentialExchangeId}" not found.`)
+      }
+
+      this.setStatus(500)
+      return apiErrorResponse(error)
+    }
+  }
+
+  /**
+   * Deletes a credential exchange record in the credential repository.
+   *
+   * @param credentialExchangeId
+   */
+  @Delete('/:credentialExchangeId')
+  public async deleteCredential(
+    @Request() request: RequestWithAgent,
+    @Path('credentialExchangeId') credentialExchangeId: RecordId,
+  ) {
+    try {
+      this.setStatus(204)
+      await request.user.agent.credentials.deleteById(credentialExchangeId)
+    } catch (error) {
+      if (error instanceof RecordNotFoundError) {
+        this.setStatus(404)
+        return apiErrorResponse(`credential exchange with id "${credentialExchangeId}" not found.`)
+      }
+
+      this.setStatus(500)
+      return apiErrorResponse(error)
+    }
+  }
+
+  /**
+   * Initiate a new credential exchange as holder by sending a propose credential message
+   * to the connection with a specified connection id.
+   *
+   * @param options
+   * @returns CredentialExchangeRecord
+   */
+  @Example<DidCommCredentialsExchangeRecord>(credentialExchangeRecordExample)
+  @Post('/propose-credential')
```

Comment:
Would it make sense to have the connection id as a path param? 

Again, I have no major preference to either API, but could be worth considering if it makes it more clear.

Reply:
However I do really like the fact to just pass the options along to credo and have the logic there. maintenance is a lot easier with this implementation.

Reply:
This is again restful best practice. We're on the `/credentials` endoint here. So passing a connectionId doesn't make sense (for the other endpoints where a credentialExchangeId is expected it does inlcude it in the path).

If we would want to pass a connectionId in path, we should add the endpoint under the connection resource:
`/didcomm/connections/<connectionId>/propose-credential`

However that gets messy as you can also do some credential operations without connectionId. So connectionId is an input parameter to create a credential exchange resource.